### ISSUE TITLE
[OTHER]: Interoperability Issue: Same C++ Class Across Different Nanobind Modules

### ISSUE THREAD

**Issue by sxsong1207** [wjakob/nanobind#807]
## Issue Description
I'm encountering type compatibility issues when trying to use the same C++ class across different nanobind modules. While the underlying C++ class is identical, Python treats the bound classes as different types.

### Global C++ Class and Binding
```cpp
// global_class.hpp
class GlobalClass {
    // definitions
};

// binding.hpp
void bind_global(nb::module_& m) {
    nb::class_<GlobalClass>(m, "GlobalClass")
        .def(nb::init<>())
        // ... other bindings
        ;
}
```

### Module A

```cpp
// module_a.cpp
NB_MODULE(nbA, m) {
    bind_global(m);
}
```

### Module B

```cpp
// module_b.cpp
NB_MODULE(nbB, m) {
    bind_global(m);
}
```

### Python calls
```python
# This fails with type error
obj = nbA.create_obj()
nbB.use_obj(obj)  # TypeError: use_obj(): incompatible function arguments.
                  # Expects nbB.GlobalClass but received nbA.GlobalClass
```

## Expected Behavior
I expect to be able to pass objects created from one module to functions in another module, since they represent the same underlying C++ class.

I noticed that your projects like Mitsuba3 and DrJit (which both use nanobind) successfully handle cross-module type compatibility. For example, you can create a vector in DrJit and process it in Mitsuba3.

## Questions
1. What is the recommended approach to achieve cross-module type compatibility with nanobind?
2. Are there specific binding patterns or configurations needed to enable this functionality? (for example, the data must be an array or simple data type)

Any guidance or examples to address this issue would be greatly appreciated!

**Comment by ThoreWietzke** [comment]
Just like in pybind, nanobind has the macro `NB_EXPORT` which makes your symbol visible. You can define it like
```c++
// global_class.hpp
class NB_EXPORT GlobalClass {
    // definitions
};
```

**Comment by sxsong1207** [comment]
Thank you for the insight! I am using GCC on Linux, so NB_EXPORT isn’t the reason. Finally, I discovered that the `FREE_THREADED` flag used when creating the target was the reason. 

My Python version is 3.12, and according to the [README](https://nanobind.readthedocs.io/en/latest/api_cmake.html#command:nanobind_add_module), this flag should be ignored. Interestingly, it causes the symbol to be separated between modules.

Once I removed that flag, I made the inter-module method calls to work.
